CONTRIBUTING.md

```diff
@@ -1,103 +1,155 @@
+     and push to the branch again (there is no need to re-create the pull request, 
+     it will automatically track modifications to your branch), e.g.:
+    ```shell
+    git add tests/test-changed-file.py
+    git commit -m "Added another test case to address reviewer feedback"
+    git push origin new-awesome-feature
+    ```
+   - Once your pull request is approved by the reviewers, it will be merged into the main codebase.
+
+## Setting up your development environment
+
+We test the library code against multiple `python` versions 
+and use [`pyenv`](https://github.com/pyenv/pyenv) to manage those. Install `pyenv` by running
 
 ```shell
-git clone git@github.com:wandb/client.git
-cd client
-pip install -e .
+curl https://pyenv.run | bash
 ```
 
-Install from pip:
+To load `pyenv` automatically, add the following lines to your shell's startup script, 
+such as `~/.bashrc` or `~/.zshrc` 
+(and then either restart the shell, run `exec $SHELL`, or `source` the changed script):
 
 ```shell
-pip install --upgrade git+git://github.com/wandb/client.git
+export PYENV_ROOT="$HOME/.pyenv"
+export PATH="$HOME/.pyenv/bin:$PATH"
+eval "$(pyenv init --path)"
+eval "$(pyenv virtualenv-init -)"
 ```
 
-Or from pypi:
+Then run the following command to set up your environment:
 
 ```shell
-pip install --upgrade wandb
+./tools/setup_dev_environment.py
 ```
 
-## Code organization
+At the first invocation, this tool will set up multiple python environments, which takes some time.
+You can set up a subset of the target environments to test against, for example:
 
-```bash
-wandb/
-├── ...
-├── apis/   # Public api (still has internal api but this should be moved to wandb/internal)
-│   ├── ...
-│   ├── internal.py
-│   ├── ...
-│   └── public.py
-├── cli/    # Handlers for command line functionality
-├── ...
-├── integration/    # Third party integration
-│   ├── fastai/
-│   ├── gym/
-│   ├── keras/
-│   ├── lightgbm/
-│   ├── metaflow/
-│   ├── prodigy/
-│   ├── sacred/
-│   ├── sagemaker/
-│   ├── sb3/
-│   ├── tensorboard/
-│   ├── tensorflow/
-│   ├── torch/
-│   ├── xgboost/
-│   └── ...
-├── ...
-├── proto/  # Protocol buffers for inter-process communication and persist file store
-├── ...
-├── sdk/    # User accessed functions [wandb.init()] and objects [WandbRun, WandbConfig, WandbSummary, WandbSettings]
-│   ├── backend/    # Support to launch internal process
-│   ├── ...
-│   ├── interface/  # Interface to backend execution
-│   ├── internal/   # Backend threads/processes
-│   └── ...
-├── ...
-├── sweeps/ # Hyperparameter sweep engine (see repo: https://github.com/wandb/sweeps)
-└── ...
+```shell
+./tools/setup_dev_environment.py -p 3.7 3.8
 ```
 
-## Setup development environment
-
-In order to run unittests please install pyenv:
+The tool will also set up [`tox`](https://github.com/tox-dev/tox), which we use 
+for automating development tasks such as code linting and testing. 
 
-(put the output in your `~/.bashrc`)
+## Building protocol buffers
```

Comment:
I think there's a step missing here - we haven't outlined the structure of the client yet so folks wont have an idea of what the internal/ user processes are. Maybe this is a good place for a very high level introduction to the strcuture/architecture of the codebase

Reply:
yea, you're right - I moved that below, but now that you bring this up, I agree that it belong here. Moved back.